packages/protocol/docs/ZKReq.md

```diff
@@ -0,0 +1,363 @@
+- `block.prevrandao` (`uint`): random number provided by the beacon chain
+- `block.timestamp` (`uint`): current block timestamp as seconds since Unix epoch
+
+We need to verify when these variables are accessed within the EVM, their values are consistent with the current world state, the block's metadata and the actual L2 block's block header:
+
+- `blockhash`: EVM allows access to the most recent 256 block hashes. All these hashes are available inside the plonk lookup table. ZKP must prove that the lookup table is consistent with L2's historical values.
+  - `blockhash(block.number - 1)`, has the same value as in the block header and is also the same value as the parent block's hash on L2.
+  - The other 255 hashes, `blockhash(block.number - 256)` to `blockhash(block.number - 2)` are checked in the anchor transaction to simplify circuits. Therefore, as long as the anchor transaction is zk-proven, these 255 ancestor hashes are proven indirectly.
+- `block.basefee`: verified to be the correct value in the anchor transaction.
+- `block.chainid`: this field is also checked by the anchor transaction, so no extra ZKP circuits are required.
+- `block.coinbase`: ZKP must verify the value must be the same as `meta.beneficiary`. Again, the metadata hash is part of the ZK instance.
+- `block.difficulty`: this is now the same as `block.prevrandao`, so we only check `block.prevrandao`.
+- `block.gaslimit`: ZKP must verify this value must equal `meta.gasLimit`.
+- `block.gasUsed`: ZKP must verify this value must be smaller than `meta.gasLimit`.
+- `block.number`: this must be checked against the block header and `meta.id`.
+- `block.prevrandao`: this must be checked against the `mixHash` field in the L2 block header and `meta.mixHash`.
+- `block.timestamp`: this must be checked against the `timestamp` field in the L2 block header and `meta.proposedAt`.
+
+### Block Header
+
+The block header is no longer presented to L1 contracts; therefore, ZKP must verify the integrity of the block header and ensure the block header can hash into the same value as `evidence.blockHash`, which is part of the ZK instance input.
+
+```solidity
+struct BlockHeader {
+  bytes32 parentHash;
+  bytes32 ommersHash;
+  address beneficiary;
+  bytes32 stateRoot;
+  bytes32 transactionsRoot;
+  bytes32 receiptsRoot;
+  bytes32[8] logsBloom;
+  uint256 difficulty;
+  uint128 height;
+  uint64 gasLimit;
+  uint64 gasUsed;
+  uint64 timestamp;
+  bytes extraData;
+  bytes32 mixHash;
+  uint64 nonce;
+  uint256 baseFeePerGas;
+  bytes32 withdrawalsRoot;
+}
+```
+
+In addition, ZKP must also prove the following:
+
+- `parentHash` must be the same as `evidence.parentHash`.
+- `ommersHash` must be the keccak256 of `[]`,or `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347`.
+- `beneficiary` must be `meta.beneficiary` (duplicated, as stated above).
+- `logsBloom` must be a `bytes32[8]` with all zeros.
+- `difficulty` must be `meta.mixHash`.
```

Comment:
```suggestion
- `difficulty` == 0.
```

As @ggkitsas has found, this should not be the `mixHash`, the `DIFFICULTY` opcode just returns the `mixHash/prevRandao`. Its value in the block header is simply `0`: https://blog.ethereum.org/2021/11/29/how-the-merge-impacts-app-layer. In the diagram below it is correct.